## Analyzing used car listing

* We have 50000 record of used car which were for sale on ebay.
* Data dictionary,
    - `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
    - `name` - Name of the car.
    - `seller` - Whether the seller is private or a dealer.
    - `offerType` - The type of listing
    - `price` - The price on the ad to sell the car.
    - `abtest` - Whether the listing is included in an A/B test.
    - `vehicleType` - The vehicle Type.
    - `yearOfRegistration` - The year in which the car was first registered.
    - `gearbox` - The transmission type.
    - `powerPS` - The power of the car in PS.
    - `model` - The car model name.
    - `kilometer` - How many kilometers the car has driven.
    - `monthOfRegistration` - The month in which the car was first registered.
    - `fuelType` - What type of fuel the car uses.
    - `brand` - The brand of the car.
    - `notRepairedDamage` - If the car has a damage which is not yet     - repaired.
    - `dateCreated` - The date on which the eBay listing was created.
    - `nrOfPictures` - The number of pictures in the ad.
    - `postalCode` - The postal code for the location of the vehicle.
    - `lastSeenOnline` - When the crawler saw this ad last online.

-------------

In [241]:
import pandas as pd
import numpy as np

In [242]:
autos = pd.read_csv('data/autos.csv', encoding='Latin-1')

In [243]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [244]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

* There are missing data in `vehicleType`, `gearbox`, `model`, `fuelType`, `nonRepairedDamage`. Most of the columns has less than 20% of missing data.
* Most of the variable has type string.
* Let's clean up column name first, according to python's convention it should be snake case.

### Clean columns

In [245]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [246]:
column_name_mapping = {'dateCrawled':'date_crawled' , 'offerType':'offer_type', 'abtest':'ab_test',
       'vehicleType':'vehicle_type', 'yearOfRegistration':'registration_year', 'gearbox':'gear_box', 'powerPS':'power_ps',
       'monthOfRegistration':'registration_month', 'fuelType':'fuel_type',
       'notRepairedDamage':'unrepaired_damage', 'dateCreated':'ad_created', 'nrOfPictures':'total_pictures', 'postalCode':'postal_code',
       'lastSeen':'last_seen'}

In [247]:
autos.rename(columns=column_name_mapping, inplace=True)

In [248]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gear_box              47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
total_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

### Initial exploration

In [249]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  total_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [250]:
autos.seller.unique()

array(['privat', 'gewerblich'], dtype=object)

In [251]:
autos.seller.value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [252]:
autos.loc[autos.seller == 'gewerblich',:]

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price  ab_test vehicle_type  registration_year  \
7738  gewerblich    Angebot  $100  control        kombi               2000   

     gear_box  power_ps   model   odometer  registration_month fuel_type  \
7738  manuell         0  megane  150,000km                   8    benzin   

        brand unrepaired_damage           ad_created  total_pictures  \
7738  renault               NaN  2016-03-15 00:00:00               0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

In [253]:
autos.offer_type.unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [254]:
autos.offer_type.value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [255]:
autos.loc[autos.offer_type == 'Gesuch', :]

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      ab_test vehicle_type  registration_year gear_box  power_ps        model  \
17541    test          bus               2005      NaN         0  transporter   

        odometer  registration_month fuel_type       brand unrepaired_damage  \
17541  150,000km                   0       NaN  volkswagen               NaN   

                ad_created  total_pictures  postal_code            last_seen  
17541  2016-04-03 00:00:00               0        29690  2016-04-05 15:16:06

In [256]:
autos.ab_test.unique()

array(['control', 'test'], dtype=object)

In [257]:
autos.ab_test.value_counts()

test       25756
control    24244
Name: ab_test, dtype: int64

In [258]:
autos.vehicle_type.unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [259]:
autos.vehicle_type.value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [260]:
autos.gear_box.unique()

array(['manuell', 'automatik', nan], dtype=object)

In [261]:
autos.gear_box.value_counts()

manuell      36993
automatik    10327
Name: gear_box, dtype: int64

In [262]:
autos.odometer.unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [263]:
autos.fuel_type.unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [264]:
autos.fuel_type.value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [265]:
autos.unrepaired_damage.unique()

array(['nein', nan, 'ja'], dtype=object)

In [266]:
autos.unrepaired_damage.value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

* Columns `seller`, `offer_type` has only 1 value deviated from others
* Columns `total_pictures` has all values 0.
* Let's drop them

In [267]:
autos = autos.drop(['total_pictures', 'seller', 'offer_type'], axis = 1)

--------------

* `price`, `odometer` has to be numerical.

In [268]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [269]:
import re

In [270]:
clean_price_regex = re.compile('[\$,]')

In [271]:
autos['price'] = (autos['price']
                      .str.replace(clean_price_regex, '')
                      .astype(int)
                 )

In [272]:
autos.price.head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

In [273]:
autos.odometer.head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [274]:
autos['odometer'] = (autos['odometer']
                     .apply(lambda x:x[:-2].replace(',',''))
                     .astype(int)
                    )

In [275]:
autos.rename(columns={'odometer':'odometer_km'}, inplace=True)

In [276]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year   gear_box  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [277]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
price                 50000 non-null int32
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gear_box              47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int32
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int32(2), int64(4), object(11)
memory usage: 6.1+ MB


### Explore `price` and `odometer`

In [281]:
autos.price.value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [284]:
autos.price.value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

* As ebay is auction site, bid generally start from $1 so we will discard all the 0 price. Furthermore, after 350000,  price is jumping rapidly, which seems odd, lets remove all price higher than 350000.

In [291]:
autos = autos.loc[autos['price'].between(1, 350000),:]

In [293]:
autos.price.describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [296]:
autos.odometer_km.value_counts().sort_index()

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

* Most of the vehicles are of highly used.

### Explore date
* Here we will explore `date_crawled`, `last_seen`, `ad_created`, `registration_month`, `registration_year` 

In [297]:
autos.date_crawled.describe()

count                   48565
unique                  46882
top       2016-03-22 09:51:06
freq                        3
Name: date_crawled, dtype: object

In [298]:
autos.last_seen.describe()

count                   48565
unique                  38474
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object

In [299]:
autos.ad_created.describe()

count                   48565
unique                     76
top       2016-04-03 00:00:00
freq                     1887
Name: ad_created, dtype: object

In [301]:
autos.registration_month.describe()

count    48565.000000
mean         5.782251
std          3.685595
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [302]:
autos.registration_year.describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

* `date_crawled`, `last_seen`, `ad_created` are represented as string

In [304]:
autos[['date_crawled', 'last_seen', 'ad_created']].head()

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

In [309]:
(autos.date_crawled
     .str[:10]
     .value_counts(normalize = True, dropna = False)
     .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

* Looks like website was crawled daily over the month of march to april 2016. Distribution seems uniform.

In [310]:
(autos.last_seen
     .str[:10]
     .value_counts(normalize = True, dropna= False)
     .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [311]:
(autos.ad_created
     .str[:10]
     .value_counts(normalize=True, dropna=False)
     .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

* Most of the ad are created within 1-2 months of listing date, oldest around 9 months.

In [312]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

* Strange!!!! car registration in year 1000 and in year 9999!!!
* Simply discard all year value less than 1900. and can can not be registered after ad is posted, so before 2016

* Let's calculate percentage of invalid data

In [313]:
(~autos.registration_year.between(1900,2016)).sum() /autos.shape[0]

0.038793369710697

* Almost 4% of observation are invalid

In [314]:
autos = autos.loc[autos.registration_year.between(1900,2016), :]

In [315]:
autos.registration_year.value_counts(normalize = True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

### Explore `brand`

In [316]:
autos.brand.unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [321]:
autos.groupby('brand').size().sort_values()

brand
lada                27
lancia              50
rover               62
trabant             65
daewoo              70
jaguar              73
saab                77
land_rover          98
subaru             100
jeep               106
daihatsu           117
dacia              123
chrysler           164
chevrolet          266
suzuki             277
porsche            286
alfa_romeo         310
kia                330
honda              366
mitsubishi         384
mini               409
volvo              427
sonstige_autos     458
hyundai            468
toyota             593
citroen            654
smart              661
mazda              709
nissan             713
skoda              766
seat               853
fiat              1197
peugeot           1393
renault           2201
ford              3263
audi              4041
mercedes_benz     4503
opel              5022
bmw               5137
volkswagen        9862
dtype: int64

* Looks like volkswagen is very popular brand.

In [326]:
autos.groupby('brand')[['price', 'odometer_km']].mean().sort_values(by = 'price')

price    odometer_km
brand                                      
daewoo           1049.000000  121642.857143
rover            1602.290323  137661.290323
daihatsu         1636.196581  116410.256410
trabant          1790.861538   54538.461538
renault          2474.864607  128071.331213
lada             2688.296296   83518.518519
fiat             2813.748538  117121.971596
opel             2975.241935  129310.035842
peugeot          3094.017229  127153.625269
saab             3211.649351  144415.584416
lancia           3376.220000  121900.000000
mitsubishi       3394.572917  126575.520833
chrysler         3465.743902  132378.048780
smart            3580.223903   99326.777610
ford             3749.469507  124266.012872
citroen          3779.139144  119694.189602
subaru           3953.290000  125300.000000
alfa_romeo       4087.690323  131338.709677
suzuki           4096.732852  108465.703971
honda            4107.857923  122172.131148
mazda            4112.596615  124464.033850
seat             4397.230950  121131.301290
nissan           4743.402525  118330.995792
volvo            4946.501171  138067.915691
toyota           5167.091062  115944.350759
hyundai          5365.254274  106442.307692
volkswagen       5402.410262  128707.158791
dacia            5915.528455   84268.292683
kia              5982.330303  112530.303030
skoda            6368.000000  110848.563969
chevrolet        6684.139098   99191.729323
bmw              8332.820518  132572.513140
mercedes_benz    8628.450366  130788.363313
audi             9336.687454  129157.386785
mini            10613.459658   88105.134474
jaguar          11635.493151  124178.082192
jeep            11650.500000  127122.641509
sonstige_autos  12338.550218   89956.331878
land_rover      19108.091837  118010.204082
porsche         45643.937063   96853.146853